test/cppsuite/tests/search_near_03.cxx

```diff
@@ -0,0 +1,288 @@
+ * Public Domain 2008-2014 WiredTiger, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "test_harness/test.h"
+#include "test_harness/util/api_const.h"
+#include "test_harness/workload/random_generator.h"
+
+using namespace test_harness;
+
+/*
+ * In this test, we want to verify search_near with prefix enabled when performing unique index
+ * insertions. During the test duration:
+ *  - N threads will keep performing unique index insertions on existing keys in the table, in which
+ * we are expecting it to fail.
+ *  - M threads traverse all the keys in all the collections, to make sure that the number of
+ * collections do not change.
+ */
+class search_near_03 : public test_harness::test {
+    std::vector<std::vector<std::string>> prefixes_map;
+    const std::string ALPHABET{"abcdefghijklmnopqrstuvwxyz"};
+
+    public:
+    search_near_03(const test_harness::test_args &args) : test(args) {}
+
+    void
+    run() override final
+    {
+        /* You can remove the call to the base class to fully customized your test. */
+        test::run();
+    }
+
+    /*
+     * A unique index has the following insertion method:
+     * 1. Insert the prefix.
+     * 2. Remove the prefix.
+     * 3. Search near for the prefix.
+     * 4. Insert the full value (prefix, id).
+     * All of these operations are wrapped in the same txn, this test attempts to test
+     * scenarios that could arise from this insertion method.
+     */
+    static bool
+    perform_unique_index_insertions(
+      thread_context *tc, scoped_cursor &cursor, collection &coll, std::string &prefix_key)
+    {
+        const char *key_tmp;
+        int exact_prefix, ret;
+
+        /* Insert the prefix. */
+        if (!tc->insert(cursor, coll.id, prefix_key))
```

Comment:
Did you encounter a situation where `insert` returned false here? Same for the `remove` below.

Reply:
Yes, actually the insert has return false the most times! And all of them returned back WT_ROLLBACK

Reply:
I wonder why...?

Reply:
I haven't looked into this full PR yet but write-conflicts perhaps?

Reply:
Just confirmed that this is most likely write conflicts. The error mostly comes from when concurrent insert threads end up on doing unique index insertions on the same existing prefix key. Thus WT_ROLLBACK gets returned cause of this.

Reply:
Ok, this makes sense then.